wdl/qc_imputation.wdl

```diff
@@ -122,53 +156,423 @@ workflow qc_imputation {
+        NR>1 {$1=$1; print $0,$h["HW_exact_p_value"]}' > tmp && mv tmp ${outname}.snpstats.txt
+
+        if [[ "${chr}" -eq 23 ]];
+        then
+            plink --allow-extra-chr --keep-allele-order --bfile ${outname} --split-x b${genome_build} no-fail --make-bed --out plink_data
+            plink --allow-extra-chr --keep-allele-order --bfile plink_data --check-sex ${sep=" " Fs} --out plink_data
+            awk '$4==2{ print $1,$2}' plink_data.sexcheck > genetic_females
+            #get non-PAR X variants in genetic females to calculate snpstats from
+            plink --allow-extra-chr --keep-allele-order --bfile plink_data --chr 23 --keep genetic_females --make-bed --out females_only
+            qctool -g females_only.bed -snp-stats -osnp females_only_chrx.snpstats.txt
+
+            # replace X chromosome hwe in snpstats based on females only
+            awk '
+            BEGIN {OFS="\t"} NR==1 {for (i=1;i<=NF;i++) h[$i]=i}
+            NR>1 && FNR==NR {
+                id=$h["rsid"];hw=$h["HW_exact_p_value"]
+                hwes[id]=hw;
+            }
+            FNR<NR && FNR == 1{ for (i=1;i<=NF;i++) h[$i]=i; $1=$1; print $0 }
+            FNR<NR && FNR > 1 {
+                id=$h["rsid"]
+                hw=$h["HW_exact_p_value"]
+                if (id in hwes) $h["HW_exact_p_value"]=hwes[id] # update HWE based on females for non-PAR variants
+                print $0
+            }
+            ' <(grep -Ev "^#" females_only_chrx.snpstats.txt) <(grep -Ev "^#" ${outname}.snpstats.txt) > tmp
+            mv tmp ${outname}".snpstats.txt"
+        fi
+
+        echo -e "`date`\tdone"
+
+    >>>
+
+    output {
+        File snpstats = outname + ".snpstats.txt"
+        File out_bed = outname + ".bed"
+        File out_bim = outname + ".bim"
+        File out_fam = outname + ".fam"
+    }
+
+    runtime {
+        docker: "${docker}"
+        memory: "24 GB"
+        cpu: 6
+        disks: "local-disk 500 HDD"
+        preemptible: 2
+        zones: "europe-west1-b europe-west1-c europe-west1-d"
+    }
+}
+
+task gather_snpstats_joint_qc {
+
+    String name
+    Array[File] snpstats
+    Array[File] exclude_variants_batch_nonpass
+    Array[File] exclude_variants_batch_missingness
+    Float af
+    Float hw
+    Float escape_hwe_maf
+    Float variant_missing_overall
+    Int non_pass_n_batches
+    Int missingness_n_batches
+    String docker
 
     command <<<
+
         set -euxo pipefail
-        echo "Execute vcf-fusion&paste&bgzip command"
-        time vcf-fusion ${sep=" " vcfs}| bgzip -@${cpus} > ${outfile}
-        tabix -p vcf ${outfile}
+
+        for file in ${sep=" " exclude_variants_batch_nonpass}; do
+            cat $file >> ${name}.exclude_upstream_variants.txt
+        done
+
+        # add variants that were non-PASS in the original data in given number of batches
+        grep batch_non_pass ${name}.exclude_upstream_variants.txt | \
+        datamash -s -g 1 count 2 | \
```

Comment:
datamash rocks..... btw. is the 2 needed in the count? Not very familiar with datamash but don't you want the number elements in the -g 1 i.e. variant id regardless of what is in the 2nd column (reason for rejection). I thought the `count 2` would get unique elements of 2nd column within a group, but it's still just counting them?

Reply:
tested out and yea the field for count is needed but feels unnecessary as the result is the same no matter what the count column is :)
anyhoo not related to this review but just discussing with you via code review as its lonely here :)

Reply:
(didn't read the second comment) You need to give a field for datamash count but it doesn't matter which one you give. countunique is separate from count and there it then matters which field you give